Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/Kconfig
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/Kconfig
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/Kconfig
@@ -69,3 +69,11 @@ config ATH11K_THERMAL
 	depends on THERMAL
 	help
 	  Enable ath11k thermal sensors and throttling support.
+
+config ATH11K_SMART_ANT_ALG
+	bool "Smart antenna algorithm support"
+	depends on ATH11k
+	default y
+	---help---
+	  This option enabled smart antenna algorithm.
+
Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/Makefile
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/Makefile
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/Makefile
@@ -26,6 +26,7 @@ ath11k-$(CPTCFG_NL80211_TESTMODE) += tes
 ath11k-$(CPTCFG_ATH11K_TRACING) += trace.o
 ath11k-$(CPTCFG_ATH11K_THERMAL) += thermal.o
 ath11k-$(CPTCFG_ATH11K_SPECTRAL) += spectral.o
+ath11k-$(CPTCFG_ATH11K_SMART_ANT_ALG) += smart_ant_api.o
 ath11k-$(CONFIG_PM) += wow.o
 
 obj-$(CPTCFG_ATH11K_AHB) += ath11k_ahb.o
Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/core.c
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/core.c
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/core.c
@@ -37,6 +37,11 @@ bool ath11k_ftm_mode;
 module_param_named(ftm_mode, ath11k_ftm_mode, bool, 0444);
 MODULE_PARM_DESC(ftm_mode, "Boots up in factory test mode");
 
+bool ath11k_enable_smart_antenna = true;
+module_param_named(enable_smart_antenna, ath11k_enable_smart_antenna, uint, 0644);
+MODULE_PARM_DESC(enable_smart_antenna, "Enable smart antenna support in fw");
+EXPORT_SYMBOL(ath11k_enable_smart_antenna);
+
 static struct ath11k_hw_params ath11k_hw_params[] = {
 	{
 		.hw_rev = ATH11K_HW_IPQ8074,
Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/core.h
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/core.h
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/core.h
@@ -32,6 +32,7 @@
 #include "spectral.h"
 #include "wow.h"
 #include "fw.h"
+#include "smart_ant.h"
 
 #define SM(_v, _f) (((_v) << _f##_LSB) & _f##_MASK)
 
@@ -651,6 +652,8 @@ struct ath11k {
 	u8 mac_addr[ETH_ALEN];
 	struct ath11k_he ar_he;
 	enum ath11k_state state;
+	struct ath11k_smart_ant_info smart_ant_info;
+	u32 rx_antenna;
 	bool supports_6ghz;
 	struct {
 		struct completion started;
Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/debug.h
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/debug.h
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/debug.h
@@ -27,6 +27,7 @@ enum ath11k_debug_mask {
 	ATH11K_DBG_DP_TX	= 0x00002000,
 	ATH11K_DBG_DP_RX	= 0x00004000,
 	ATH11K_DBG_CE		= 0x00008000,
+	ATH11K_DBG_SMART_ANT	= 0x00100000,
 };
 
 static inline const char *ath11k_dbg_str(enum ath11k_debug_mask mask)
Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/hal_rx.h
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/hal_rx.h
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/hal_rx.h
@@ -191,6 +191,7 @@ struct hal_rx_mon_ppdu_info {
 	u8 rxpcu_filter_pass;
 	char rssi_chain[8][8];
 	struct hal_rx_user_status userstats;
+	u32 rx_antenna;
 };
 
 #define HAL_RX_PPDU_START_INFO0_PPDU_ID		GENMASK(15, 0)
Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/mac.c
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/mac.c
@@ -7467,6 +7467,8 @@ ath11k_mac_update_vif_chan(struct ath11k
 			continue;
 		}
 
+		ath11k_smart_ant_disable(arvif);
+
 		ret = ath11k_mac_setup_bcn_tmpl(arvif);
 		if (ret)
 			ath11k_warn(ab, "failed to update bcn tmpl during csa: %d\n",
@@ -8095,6 +8097,19 @@ ath11k_mac_op_assign_vif_chanctx(struct
 		}
 
 		arvif->is_started = true;
+
+		ret = ath11k_smart_ant_enable(arvif);
+		if (ret) {
+			ath11k_warn(ab, "failed to enable smart antenna algorithm %d\n", ret);
+			goto out;
+		}
+
+		ret = ath11k_smart_ant_set_default(arvif);
+		if (ret) {
+			ath11k_warn(ab, "failed to set default smart antenna configuration %d\n", ret);
+			ath11k_smart_ant_disable(arvif);
+			goto out;
+		}
 	}
 
 	if (arvif->vdev_type != WMI_VDEV_TYPE_MONITOR &&
Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/wmi.c
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/wmi.c
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/wmi.c
@@ -161,6 +161,8 @@ static const struct wmi_tlv_policy wmi_t
 		.min_len = sizeof(struct ath11k_wmi_p2p_noa_info) },
 	[WMI_TAG_P2P_NOA_EVENT] = {
 		.min_len = sizeof(struct wmi_p2p_noa_event) },
+	[WMI_TAG_PEER_RATECODE_LIST_EVENT] = {
+		.min_len = sizeof(struct wmi_peer_ratecode_list_fixed_param) },
 };
 
 #define PRIMAP(_hw_mode_) \
@@ -4186,6 +4188,7 @@ ath11k_wmi_copy_resource_config(struct w
 	wmi_cfg->flags2 = WMI_RSRC_CFG_FLAG2_CALC_NEXT_DTIM_COUNT_SET;
 	wmi_cfg->ema_max_vap_cnt = tg_cfg->ema_max_vap_cnt;
 	wmi_cfg->ema_max_profile_period = tg_cfg->ema_max_profile_period;
+	wmi_cfg->smart_ant_cap = 1;
 }
 
 static int ath11k_init_cmd_send(struct ath11k_pdev_wmi *wmi,
@@ -8908,6 +8911,10 @@ static void ath11k_wmi_tlv_op_rx(struct
 	case WMI_P2P_NOA_EVENTID:
 		ath11k_wmi_p2p_noa_event(ab, skb);
 		break;
+	case WMI_PEER_RATECODE_LIST_EVENTID:
+		ath11k_wmi_event_peer_ratecode_list(ab, skb);
+		break;
+
 	default:
 		ath11k_dbg(ab, ATH11K_DBG_WMI, "unsupported event id 0x%x\n", id);
 		break;
Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/wmi.h
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/wmi.h
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/wmi.h
@@ -12,6 +12,10 @@
 
 struct ath11k_base;
 struct ath11k;
+struct ath11k_vif;
+struct ath11k_smart_ant_info;
+struct ath11k_smart_ant_node_config_params;
+struct ath11k_smart_ant_train_info;
 struct ath11k_fw_stats;
 struct ath11k_fw_dbglog;
 struct ath11k_vif;
@@ -6358,6 +6362,101 @@ enum wmi_sta_keepalive_method {
 #define WMI_STA_KEEPALIVE_INTERVAL_DEFAULT	30
 #define WMI_STA_KEEPALIVE_INTERVAL_DISABLE	0
 
+/*Smart antenna related defs*/
+#define WMI_SMART_ANT_MODE_SERIAL 0
+#define WMI_SMART_ANT_MODE_PARALLEL 1
+#define WMI_SMART_ANTENNA_HAL_MAX 4
+#define WMI_SMART_ANT_MAX_SERIAL_ANTENNA 2
+#define WMI_SMART_MAX_RATE_SERIES 2
+
+#define WMI_CCK_OFDM_RATES_MAX 12
+#define WMI_MCS_RATES_MAX 96
+#define WMI_RATE_COUNT_MAX 4
+
+struct wmi_peer_ratecode_list_fixed_param {
+	struct wmi_mac_addr macaddr;
+	u32 ratecount;
+	u32 vdev_id;
+	u32 pdev_id;
+} __packed;
+
+struct wmi_peer_cck_ofdm_rate_info {
+	u32 ratecode_legacy;
+} __packed;
+
+struct wmi_peer_mcs_rate_info {
+	u32 rt_code_20;
+	u32 rt_code_40;
+	u32 rt_code_80;
+} __packed;
+
+struct wmi_pdev_set_smart_ant_cmd {
+	u32 tlv_header;
+	union {
+		u32 mac_id;
+		u32 pdev_id;
+	};
+	u32 enable;
+	u32 mode;
+	u32 rx_antenna;
+	u32 tx_default_antenna;
+} __packed;
+
+struct wmi_pdev_smart_ant_gpio_handle_cmd {
+	u32 tlv_header;
+	u32 gpio_pin;
+	u32 gpio_func;
+	u32 pdev_id;
+} __packed;
+
+struct wmi_pdev_set_rx_antenna_cmd {
+	u32 tlv_header;
+	union {
+		u32 mac_id;
+		u32 pdev_id;
+	};
+	u32 rx_antenna;
+} __packed;
+
+struct wmi_pdev_set_tx_antenna_cmd {
+	u32 tlv_header;
+	u32 vdev_id;
+	struct wmi_mac_addr macaddr;
+} __packed;
+
+struct wmi_peer_set_smart_tx_ant_series_cmd {
+	u32 tlv_header;
+	u32 ant_series;
+} __packed;
+
+struct wmi_peer_set_smart_ant_train_ant_fixed_param_cmd {
+	u32 tlv_header;
+	u32 vdev_id;
+	struct wmi_mac_addr macaddr;
+	u32 numpkts;
+} __packed;
+
+struct wmi_peer_set_smart_ant_train_ant_param {
+	u32 tlv_header;
+	union {
+		u32 train_rate_series;
+		u32 train_rate_series_lo;
+	};
+	u32 train_antenna_series;
+	/* Rate control flags for future use */
+	u32 rc_flags;
+	u32 train_rate_series_hi;
+} __packed;
+
+struct wmi_peer_set_smart_ant_node_config_ops_cmd {
+	u32 tlv_header;
+	u32 vdev_id;
+	struct wmi_mac_addr mac_addr;
+	u32 cmd_id;
+	u32 args_count;
+} __packed;
+
+
 enum wmi_wmm_params_type {
 	WMI_WMM_PARAM_TYPE_LEGACY = 0,
 	WMI_WMM_PARAM_TYPE_11AX_MU_EDCA = 1,
@@ -6554,5 +6653,20 @@ bool ath11k_wmi_supports_6ghz_cc_ext(str
 int ath11k_wmi_send_vdev_set_tpc_power(struct ath11k *ar,
 				       u32 vdev_id,
 				       struct ath11k_reg_tpc_power_info *param);
+int ath11k_wmi_pdev_enable_smart_ant(struct ath11k *ar,
+				     struct ath11k_smart_ant_info *info);
+int ath11k_wmi_pdev_disable_smart_ant(struct ath11k *ar,
+				      struct ath11k_smart_ant_info *info);
+int ath11k_wmi_peer_set_smart_tx_ant(struct ath11k *ar, u32 vdev_id,
+				     const u8 *macaddr, const u32 *tx_antenna);
+int ath11k_wmi_pdev_set_rx_ant(struct ath11k *ar, u32 antenna);
+int ath11k_wmi_peer_set_smart_ant_node_config(struct ath11k *ar, u8 *mac,
+		struct ath11k_smart_ant_node_config_params *param);
+int ath11k_wmi_peer_set_smart_ant_train_info(struct ath11k *ar, u32 vdev_id,
+					 u8 *mac,
+					 struct ath11k_smart_ant_train_info *param);
+int ath11k_wmi_peer_set_smart_ant_train_ant_param_cmd(struct ath11k *ar,
+						  u32 vdev_id, const u8 *macaddr,
+						  const struct wmi_peer_set_smart_ant_train_ant_fixed_param_cmd *arg);
 
 #endif
Index: backports-6.12.6/drivers/net/wireless/ath/ath11k/hw.c
===================================================================
--- backports-6.12.6.orig/drivers/net/wireless/ath/ath11k/hw.c
+++ backports-6.12.6/drivers/net/wireless/ath/ath11k/hw.c
@@ -205,6 +205,7 @@ static void ath11k_init_wmi_config_ipq80
 	config->ema_max_vap_cnt = ab->num_radios;
 	config->ema_max_profile_period = TARGET_EMA_MAX_PROFILE_PERIOD;
 	config->beacon_tx_offload_max_vdev += config->ema_max_vap_cnt;
+	config->smart_ant_cap = 1;
 }
 
 static int ath11k_hw_mac_id_to_pdev_id_ipq8074(struct ath11k_hw_params *hw,
