--- a/drivers/net/wireless/ath/ath11k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_tx.c
@@ -775,6 +775,15 @@ void ath11k_dp_tx_completion_handler(str
 			wake_up(&ar->dp.tx_empty_waitq);
 		ath11k_dp_tx_complete_msdu(ar, msdu, &ts);
+		mac_id_scheduled[mac_id] = true;
+	}
+
+	/* Schedule remaining Tx packets for the radio */
+	for (i = 0; i < MAX_RADIOS; i++) {
+		if (!mac_id_scheduled[i])
+			continue;
+		ar = ab->pdevs[i].ar;
+		ath11k_mac_tx_push_pending(ar);
	}
 }
